### ISSUE TITLE
Issue with fit_all_trajectories

### ISSUE THREAD

**Issue by craiganderson89** [hafen/hbgd#6]
I received errors for some models when running the fit_all_trajectories model on the PROVIDE dataset.

The following models are affected:

fda
gam
rlm
sitar
smooth.spline

The other models appear to work properly.

The error traces back to line 223 of the script, and gives the error _"attempt to apply transformation failed with error: Error in abs(z) : non-numeric argument to mathematical function"._

This code uses distributed data etc and I don't fully understand how it works, so I'm not sure how to fix it.

**Comment by hafen** [comment]
Thanks - looking into it.

**Comment by hafen** [comment]
Can you reinstall (from v0.3 branch) and try again?  I have pushed a fix but haven't tested against all methods.

**Comment by craiganderson89** [comment]
I've rerun everything on provide, and it seems to all be working with the exception of the SITAR model now.  I'm not sure if it's relevant that SITAR is the only model that uses the unconverted height and weight rather than Z-scores?

**Comment by hafen** [comment]
Yes good catch.  Can you reinstall and try SITAR again?

**Comment by craiganderson89** [comment]
I am still receiving a similar error for SITAR on the same line of code.  The error refers to "subjid=1270", which is a child with a single observation.

**Comment by hafen** [comment]
Strange.  It works for me with subject 1270.  Can you make sure you have the latest installed?

**Comment by hafen** [comment]
Note that you'll also have to re-fit SITAR.  Might be best to start with a new session.

**Comment by craiganderson89** [comment]
Is the most recent version "Commit: 9fef76b..."?

I am running the model on htcm with random holdout and I am still receiving that error message.

**Comment by hafen** [comment]
Yes that's the one.  I didn't do random holdout though.

Here's the code I used:

``` r
dat <- fix_height(dat)
sitmod <- get_fit(dat, y_var = "htcm", method = "sitar")
tmp <- subset(dat, subjid == 1270)
fit_trajectory(tmp, sitmod)
```

If that same code gives you an error, could you send the result of `traceback()` right after the error?  This will help figure out where it occurred without dealing with the division stuff in `fit_all_trajectories()`

**Comment by craiganderson89** [comment]
That worked for me, so I think it definitely is the holdout which is the issue.

I was having a play around with the code, and I think the error comes when we have holdout=TRUE in the model, but a child with just one observation.

On line 106 of analysis_fit_traj, there is a section of code which says:

  if(holdout) {
    res$holdout <- data.frame(x = x[dat2$hold], y = y[dat2$hold])
    res$xy$hold <- dat2$hold
  }

At this stage, the object res$holdout is a matrix with 0 rows and 2 columns, and I believe that causes problems later on at line 185 when we attempt to do a transformation.

I'm not exactly sure what the best way to fix it is - would it affect anything else if we change line 185 to:

if(!is.null(res$holdout) & nrow(res$holdout)>0).

**Comment by hafen** [comment]
Can you check the holdout indicator variable for subject 1270?  In the raw data there are 2 records but one is NA.  I ran the following without error:

``` r
dat <- fix_height(dat)
dat <- subset(dat, !is.na(dat$htcm))
dat <- add_holdout_ind(dat)
sitmod <- get_fit(dat, y_var = "htcm", method = "sitar")
tmp <- subset(dat, subjid == 1270)
fit_trajectory(tmp, sitmod)
```

**Comment by hafen** [comment]
Oops I didn't specify `holdout = TRUE` in `get_fit()`.  I think you're right.  Let me try again.

**Comment by hafen** [comment]
You're suggestion is correct.  Thanks.  I just pushed the change.

**Comment by craiganderson89** [comment]
Works perfectly now - thanks!

**Comment by craiganderson89** [comment]
I have been trying to re-run the updated version of this code, with the fix made for the subject-specific methods.  There's now an error when fitting these methods for random holdout.

The error occurs when the randomly held out datapoint is the first datapoint for the a given child (ie hold[1]=TRUE).

The traceback error I received was as follows

Error: subjid=1107: attempt to apply transformation failed with error: Error in getbasismatrix(evalarg, basisobj, nderiv, returnMatrix) : 
  as.numeric(evalarg) contains 1 NA;  class(evalarg) = integer 
17 stop(keyStr, ": attempt to apply transformation failed with error: ", 
    geterrmessage(), call. = FALSE) 
16 applyTransform(curTrans, list(map.keys[[i]], map.values[[i]]), 
    env = environment()) at <text>#6
15 eval(expr, envir, enclos) 
14 eval(map, envir = mapEnv) 
13 mrExecInternal.kvMemoryList(data, setup = setup, map = map, reduce = reduce, 
    output = output, control = control, params = params) 
12 mrExecInternal(data, setup = setup, map = map, reduce = reduce, 
    output = output, control = control, params = params) 
11 mrExec(data, map = map, reduce = reduce, output = output, overwrite = overwrite, 
    params = c(parList, params), packages = c(globalVarList$packages, 
        packages), control = control) 
10 recombine(x, combine = combDdo(), output = output, overwrite = overwrite, 
    control = control) 
9 drPersist(.) 
8 function_list[[i]](value) 
7 freduce(value, `_function_list`) 
6 `_fseq`(`_lhs`) 
5 eval(expr, envir, enclos) 
4 eval(quote(`_fseq`(`_lhs`)), env, env) 
3 withVisible(eval(quote(`_fseq`(`_lhs`)), env, env)) 
2 trans_dat %>% drPersist() %>% drFilter(function(x) length(x) != 
    0) at analysis_fit_traj.R#231
1 fit_all_trajectories(data_fit, mod_fit) 
In addition: Warning message:
In eval.fd(newdata, object$fd, Lfdobj, returnMatrix = returnMatrix) :
  Values in argument 'evalarg' are outside  of permitted range and will be ignored.

**Comment by hafen** [comment]
Interesting.  I didn't experience this but must have not had subjects with first as holdout.  I think this is coming from fda and I think it is because fda doesn't predict outside of the bounds of the data it was fit on.  For now, I'd ignore fda and just do the others while I look into how to deal with this.

**Comment by craiganderson89** [comment]
I will do - thanks.

**Comment by craiganderson89** [comment]
I've downloaded the newest version, which no longer gives an error for the fda - thanks for updating.

It still produces NAs in cases where the holdout are at the beginning or end, but I assume you know that this is the case.  Is this likely to be fixed in the long term, or is it difficult/impossible to fit?

**Comment by hafen** [comment]
I think this is impossible to fix for fda unfortunately and therefore it should not be used for holdout MSE computations as it will bias results toward making it look better than it is.

**Comment by craiganderson89** [comment]
That's fine - I assumed that might be the case given the nature of the model.
